File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> mp;
        int prefix_sum = 0, max_length = 0;

        mp[0] = -1; // Handle case where the subarray starts from index 0

        for (int i = 0; i < nums.size(); i++) {
            prefix_sum += (nums[i] == 0) ? -1 : 1; // Convert 0s to -1

            if (mp.find(prefix_sum) != mp.end()) {
                max_length = max(max_length, i - mp[prefix_sum]);  
            } else {
                mp[prefix_sum] = i; // Store first occurrence
            }
        }
        return max_length;
    }
};

File: 0525-contiguous-array/0525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        HashMap<Integer, Integer> prefixMap = new HashMap<>();
        int prefixSum = 0;
        int ans = 0;
        prefixMap.put(0,-1); // store index cuz we need max len 
        for (int i = 0; i < nums.length; i++){
            prefixSum += nums[i] == 0 ? -1 : 1;
            if (prefixMap.containsKey(prefixSum)){
                int prevIndex = prefixMap.get(prefixSum);
                ans = Math.max(ans, i-prevIndex); // max length
            } else {
                // only put if its not in the map since we need earliest index to get max len
                prefixMap.put(prefixSum, i);
            }
        }
        return ans;
    }
}